package/piksi_system_daemon/src/main.c

```diff
@@ -199,18 +199,51 @@ static int date_string_get(const char *timestamp_string,
 static void img_tbl_settings_setup(settings_ctx_t *settings_ctx)
 {
   char name_string[64];
+  char uimage_string[64];
+  bool is_dev = false;
   if (file_read_string("/img_tbl/boot/name",
                        name_string, sizeof(name_string)) == 0) {
-    /* firmware_build_id contains the full image name */
+
+    if (file_read_string("/uimage_ver/name",
+                         uimage_string, sizeof(uimage_string)) != 0) {
+      sprintf(uimage_string, "unknown_uimage_version");
+    }
+
+    static char imageset_build_id[sizeof(name_string)];
+    strncpy(imageset_build_id, name_string, sizeof(imageset_build_id));
+    settings_register_readonly(settings_ctx, "system_info", "imageset_build_id",
+                               imageset_build_id, sizeof(imageset_build_id),
+                               SETTINGS_TYPE_STRING);
+
+    /* If image_table version starts with DEV, we booted a DEV build
+     * If we have DEV:
+     *   - firmware_build_id, firwmare_version, firmware_build_date come from uimage version
+     *   - "imageset_build_id" always says what imageset was used
+     * If we have PROD
+         - imageset_build_id and firmware_build_id should be identical
+         - we use the firmware_build_id and firmware_build_date from the image_set as they are
+           more relevant for production builds and matches legacy behavior
+     */
+
     static char firmware_build_id[sizeof(name_string)];
-    strncpy(firmware_build_id, name_string, sizeof(firmware_build_id));
+    if (strncmp(name_string, "DEV", 3) == 0) {
```

Comment:
The logic here seems like it can be simplified a little:
```
strncpy(firmware_build_id, uimage_string, sizeof(firmware_build_id));
is_dev = strncmp(name_string, "DEV", 3) == 0;
if (!is_dev) {
    if (strncmp(imageset_build_id, uimage_string, sizeof(imageset_build_id)) != 0) {
        /* we should never get here, in PROD, imageset_version and uimage version should match every time*/
        piksi_log(LOG_ERR, "Production build detected where uimage_build_id != imageset_build_id");
    }
}
```